### PR TITLE
Rule S4142 Duplicate values should not be passed as arguments (SONARV…

### PR DESCRIPTION
**Pull request by valhristov**
…B-243)

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+namespace SonarAnalyzer.Rules
+{
+    public abstract class DoNotPassSameValueAsMultipleArgumentsBase : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4142";
+        internal const string MessageFormat =
+            "Verify that this is the intended value; it is the same as the {0} argument.";
+    }
+
+    public abstract class DoNotPassSameValueAsMultipleArgumentsBase<TSyntaxKind, TInvocationSyntax, TArgumentSyntax>
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs]: MAJOR Reduce the number of generic parameters in the 'DoNotPassSameValueAsMultipleArgumentsBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs]: MAJOR Reduce the number of generic parameters in the 'DoNotPassSameValueAsMultipleArgumentsBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotPassSameValueAsMultipleArguments.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+        protected override DiagnosticDescriptor Rule => rule;
+
+        protected override GeneratedCodeRecognizer GeneratedCodeRecognizer =>
+            Helpers.VisualBasic.GeneratedCodeRecognizer.Instance;
+
+        protected override SyntaxKind InvocationExpressionSyntaxKind =>
+            SyntaxKind.InvocationExpression;
+
+        protected override SeparatedSyntaxList<ArgumentSyntax> GetArguments(InvocationExpressionSyntax invocation) =>
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Rules/DoNotPassSameValueAsMultipleArguments.cs]: Nitpicking: you could order last 3 methods the same way in VB and C#

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class DoNotPassSameValueAsMultipleArgumentsBase : SonarDiagnosticAnalyzer
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs]: Let's try to define a pattern and stick to it. You create a base class (as it was done previously) but @michalb-sonar put everything in the same class.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotPassSameValueAsMultipleArguments.vb [hunk 4]
```diff
@@ -0,0 +1,84 @@
+
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotPassSameValueAsMultipleArguments.vb]: Why those extra empty lines?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        internal const string DiagnosticId = "S4142";
```
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs]: Do we really need them `internal`?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs (hunk 1 above)
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs]: MAJOR Reduce the number of generic parameters in the 'DoNotPassSameValueAsMultipleArgumentsBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotPassSameValueAsMultipleArguments.vb (hunk 4 above)
Comment (valhristov) [reply sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotPassSameValueAsMultipleArguments.vb]: Telerik...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+namespace SonarAnalyzer.Rules
+{
+    public abstract class DoNotPassSameValueAsMultipleArgumentsBase : SonarDiagnosticAnalyzer
+    {
+        protected const string DiagnosticId = "S4142";
+        protected const string MessageFormat =
+            "Verify that this is the intended value; it is the same as the {0} argument.";
+    }
+
+    public abstract class DoNotPassSameValueAsMultipleArgumentsBase<TSyntaxKind, TInvocationSyntax, TArgumentSyntax>
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/DoNotPassSameValueAsMultipleArgumentsBase.cs]: MAJOR Reduce the number of generic parameters in the 'DoNotPassSameValueAsMultipleArgumentsBase' class to no more than the 2 authorized. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS2436)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

**Pull request merged**